### PR TITLE
fix metric calculation with multiple GPUs for semantic segmentation

### LINKED ISSUE #165: val/miou_best metric goes wrong when training with multiple GPU
Hi Damien,

Thanks for your great work!

image

Unfortunately, there seems to be a bug after you have changed some code in batch data processing. The val/miou and val/imou_best are not consistent when training with multiple GPUs and the latter is wrong. Could you have a look and fix this issue? 

Hope you have a good day!

Best regards,
Shanci

### PR DESCRIPTION
**Pull request by Shanci-Li**
## What does this PR do?

This PR gathers the confusion matrix computed on the unique mini-batch assigned to each GPU when training with DDP and then calculates the metrics based on the updated confusion matrix. 

Fixes #165 

## Before submitting

- [ ] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **test your PR locally** with `pytest` command?
- [ ] Did you **run pre-commit hooks** with `pre-commit run -a` command?

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/models/semantic.py [hunk 1]
```diff
@@ -700,33 +710,42 @@ def validation_step_log_metrics(self) -> None:
...
-        self.log("val/oa", oa, prog_bar=True)
-        self.log("val/macc", macc, prog_bar=True)
-        for iou, seen, name in zip(*self.val_cm.iou(), self.class_names):
-            if seen:
-                self.log(f"val/iou_{name}", iou, prog_bar=True)
+        if self.trainer.num_devices > 1:
+            epoch_cm = torch.sum(self.all_gather(self.val_cm.confmat), dim=0)
+        else:
+            epoch_cm = self.val_cm
+        # Reset metrics accumulated over the last epoch
```
Comment (drprojects) [inline src/models/semantic.py:719]: I thin the call to `.reset()` should happen last. You should compute and log your epoch metrics first and then log them. Maybe in DDP it will work because you copied the content of the metrics object from each process (to be verified) when calling `all_gather()` but I fear it won't work in the single-process case. Just to be sure, please move the `.reset()` calls to the end.

File: src/models/semantic.py [hunk 2]
```diff
@@ -700,33 +710,42 @@ def validation_step_log_metrics(self) -> None:
...
+        self.val_cm.reset()
+
```
Comment (drprojects) [inline src/models/semantic.py:721]: I thin the call to `.reset()` should happen last. You should compute and log your epoch metrics first and then log them. Maybe in DDP it will work because you copied the content of the metrics object from each process (to be verified) when calling `all_gather()` but I fear it won't work in the single-process case. Just to be sure, please move the `.reset()` calls to the end.

File: src/models/semantic.py [hunk 3]
```diff
@@ -817,22 +836,30 @@ def on_test_epoch_end(self) -> None:
...
 
         # Log confusion matrix to wandb
         if isinstance(self.logger, WandbLogger):
             self.logger.experiment.log({
                 "test/cm": wandb_confusion_matrix(
-                    self.test_cm.confmat, class_names=self.class_names)})
+                    epoch_cm.confmat, class_names=self.class_names)})
 
-        # Reset metrics accumulated over the last epoch
-        self.test_cm.reset()
```
Comment (drprojects) [inline src/models/semantic.py:835]: Why remove the call to .reset() ? Have you tested this still preserved the single-device behavior and metrics ?

File: src/models/semantic.py [hunk 4]
```diff
@@ -700,33 +710,42 @@ def validation_step_log_metrics(self) -> None:
...
-            if seen:
-                self.log(f"val/iou_{name}", iou, prog_bar=True)
+        if self.trainer.num_devices > 1:
+            epoch_cm = torch.sum(self.all_gather(self.val_cm.confmat), dim=0)
+        else:
+            epoch_cm = self.val_cm
+        # Reset metrics accumulated over the last epoch
+        self.val_cm.reset()
+
+        epoch_cm = ConfusionMatrix(self.num_classes).from_confusion_matrix(epoch_cm)
```
Comment (drprojects) [inline src/models/semantic.py:722]: This is multi-device-specific. Please move in the if statement after you gather the confmats from each device

File: src/models/semantic.py [hunk 5]
```diff
@@ -817,22 +836,30 @@ def on_test_epoch_end(self) -> None:
...
             return
 
+        if self.trainer.num_devices > 1:
+            epoch_cm = torch.sum(self.all_gather(self.test_cm.confmat), dim=0)
+        else:
+            epoch_cm = self.test_cm
+        # Reset metrics accumulated over the last epoch
+        self.test_cm.reset()
+
+        epoch_cm = ConfusionMatrix(self.num_classes).from_confusion_matrix(epoch_cm)
```
Comment (drprojects) [inline src/models/semantic.py:846]: This is multi-device-specific. Please move in the if statement after you gather the confmats from each device

File: src/models/semantic.py [hunk 6]
```diff
@@ -817,22 +836,30 @@ def on_test_epoch_end(self) -> None:
             self.test_cm.reset()
             return
 
+        if self.trainer.num_devices > 1:
+            epoch_cm = torch.sum(self.all_gather(self.test_cm.confmat), dim=0)
+        else:
+            epoch_cm = self.test_cm
+        # Reset metrics accumulated over the last epoch
+        self.test_cm.reset()
```
Comment (drprojects) [inline src/models/semantic.py:844]: I think the call to .reset() should happen last. You should compute and log your epoch metrics first and then log them. Maybe in DDP it will work because you copied the content of the metrics object from each process (to be verified) when calling all_gather() but I fear it won't work in the single-process case. Just to be sure, please move the .reset() calls to the end.

**Review by drprojects — COMMENTED** [review]
Thanks for this PR ! I went through your code and I have concerns regarding the order in which you gather the metrics from each process and the calls to `.reset()`. I have not tested your modifications on a machine but I suspect some of these changes break the single-device behavior. Have you checked the behavior and metrics are still the same in the single-device scenario ? Please see my comments in the code for more details.

Also, if you feel capable, would you mind having a shot at the distributed panoptic segmentation metrics too (ie `src/models/semantic.py`) ? I did not look much into it but presume you should be able to accumulate the internal states from `PanopticQuality3D` and `MeanAveragePrecision3D` objects from multiple devices by leveraging the `.update()` method of each of these classes (see their respective implems to understand what is happening under the hood).

**Comment by Shanci-Li** [comment]
Yes, you are right! I did not pay attention to the influence of my code on single-device training. My project requires vectorization and modeling after the semantic segmentation and I moved to this task once I got an acceptable model. I have pushed a new commit that fixes these concerns.

For the inspection of panoptic segmentation, I would like to help. But I am not sure when I have time to work on that ... I will try to do it asap.

**Comment by drprojects** [comment]
Super, thanks for your prompt modifications ! Be sure I really appreciate the help !

Have you tested on your machine for the single-gpu scenario ? Did you check whether anything breaks and whether the metrics are coherent before and after the modifications ?

Regarding the panoptic segmentation metrics, may I leave this PR open for now to give you a bit of time to have a shot at it ?

**Comment by Shanci-Li** [comment]
I tested the training in DDP and single-gpu scenario last night and the training works well. Now the metrics are logged coherently in both cases. With the default configuration file on KITTI-360 dataset (200 epoch), the single-gpu achieves 54.43 best val miou while DDP stops at 52.11. I suppose this is casued by the difference of batch size and learning rate. But I am sure that the error of inconsistence between val_miou and best_val_miou no longer exsists.

For the inspection of panoptic segmentation, I suggest to close this PR if you do not mind. Once I finish the modification I will launch a new one. As so far I have no guarantee I could finish that in a short time.

**Review by drprojects — APPROVED** [review]
